x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheServiceTests.java

```diff
@@ -0,0 +1,165 @@
+import org.elasticsearch.common.io.PathUtils;
+import org.elasticsearch.common.io.PathUtilsForTesting;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.store.cache.CacheFile;
+import org.elasticsearch.index.store.cache.CacheKey;
+import org.elasticsearch.index.store.cache.TestUtils.FSyncTrackingFileSystemProvider;
+import org.elasticsearch.repositories.IndexId;
+import org.elasticsearch.snapshots.SnapshotId;
+import org.elasticsearch.xpack.searchablesnapshots.AbstractSearchableSnapshotsTestCase;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.SortedSet;
+
+import static org.elasticsearch.index.store.cache.TestUtils.randomPopulateAndReads;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+
+public class CacheServiceTests extends AbstractSearchableSnapshotsTestCase {
+
+    private static FSyncTrackingFileSystemProvider fileSystemProvider;
+
+    @BeforeClass
+    public static void installFileSystem() {
+        fileSystemProvider = new FSyncTrackingFileSystemProvider(PathUtils.getDefaultFileSystem(), createTempDir());
+        PathUtilsForTesting.installMock(fileSystemProvider.getFileSystem(null));
+    }
+
+    @AfterClass
+    public static void removeFileSystem() {
+        PathUtilsForTesting.teardown();
+    }
+
+    public void testCacheSynchronization() throws Exception {
+        final int numShards = randomIntBetween(1, 3);
+        final Index index = new Index(randomAlphaOfLength(5).toLowerCase(Locale.ROOT), UUIDs.randomBase64UUID(random()));
+        final SnapshotId snapshotId = new SnapshotId("_snapshot_name", UUIDs.randomBase64UUID(random()));
+        final IndexId indexId = new IndexId("_index_name", UUIDs.randomBase64UUID(random()));
+
+        logger.debug("--> creating shard cache directories on disk");
+        final Path[] shardsCacheDirs = new Path[numShards];
+        for (int i = 0; i < numShards; i++) {
+            final Path shardDataPath = randomFrom(nodeEnvironment.availableShardPaths(new ShardId(index, i)));
+            assertFalse(Files.exists(shardDataPath));
+
+            logger.debug("--> creating directories [{}] for shard [{}]", shardDataPath.toAbsolutePath(), i);
+            shardsCacheDirs[i] = Files.createDirectories(CacheService.resolveSnapshotCache(shardDataPath).resolve(snapshotId.getUUID()));
+        }
+
+        try (CacheService cacheService = defaultCacheService()) {
+            cacheService.start();
+
+            logger.debug("--> setting large cache sync interval (explicit cache synchronization calls in test)");
+            cacheService.setCacheSyncInterval(TimeValue.timeValueMillis(Long.MAX_VALUE));
```

Comment:
I think we should do this before `cacheService.start()` to avoid a race condition of the schedule task "just started to run" before we change this, which could interfere with the testing below.

Reply:
Right - I pushed 6066924795a61fb689cf0b2740e17074019b5eeb